### PR TITLE
[Alerting] Automatic request instrumentation

### PR DESCRIPTION
**Pull request by owen-d**
This PR does some refactoring and enhances our code generation to automatically instrument histograms for the unified alerting API. Some of this is based upon https://github.com/weaveworks/common and extended here. I've also included some extra metrics for later use in the evaluation and notification engine.

As an added benefit, it moves some of the complexity into the metrics file and hooks into this via code generation so the handwritten implementations shouldn't need be aware of instrumentation concerns.

It injects instrumentation like
```
# HELP grafana_alerting_request_duration_seconds Histogram of requests to the Alerting API
# TYPE grafana_alerting_request_duration_seconds histogram
grafana_alerting_request_duration_seconds_bucket{backend="grafana",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="202",le="0.005"} 1
grafana_alerting_request_duration_seconds_bucket{backend="grafana",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="202",le="0.01"} 1
grafana_alerting_request_duration_seconds_bucket{backend="grafana",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="202",le="0.025"} 1
grafana_alerting_request_duration_seconds_bucket{backend="grafana",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="202",le="0.05"} 1
grafana_alerting_request_duration_seconds_bucket{backend="grafana",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="202",le="0.1"} 1
grafana_alerting_request_duration_seconds_bucket{backend="grafana",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="202",le="0.25"} 1
grafana_alerting_request_duration_seconds_bucket{backend="grafana",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="202",le="0.5"} 1
grafana_alerting_request_duration_seconds_bucket{backend="grafana",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="202",le="1"} 1
grafana_alerting_request_duration_seconds_bucket{backend="grafana",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="202",le="2.5"} 1
grafana_alerting_request_duration_seconds_bucket{backend="grafana",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="202",le="5"} 1
grafana_alerting_request_duration_seconds_bucket{backend="grafana",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="202",le="10"} 1
grafana_alerting_request_duration_seconds_bucket{backend="grafana",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="202",le="+Inf"} 1
grafana_alerting_request_duration_seconds_sum{backend="grafana",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="202"} 0.000327971
grafana_alerting_request_duration_seconds_count{backend="grafana",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="202"} 1
grafana_alerting_request_duration_seconds_bucket{backend="proxy",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="200",le="0.005"} 0
grafana_alerting_request_duration_seconds_bucket{backend="proxy",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="200",le="0.01"} 0
grafana_alerting_request_duration_seconds_bucket{backend="proxy",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="200",le="0.025"} 0
grafana_alerting_request_duration_seconds_bucket{backend="proxy",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="200",le="0.05"} 0
grafana_alerting_request_duration_seconds_bucket{backend="proxy",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="200",le="0.1"} 0
grafana_alerting_request_duration_seconds_bucket{backend="proxy",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="200",le="0.25"} 1
grafana_alerting_request_duration_seconds_bucket{backend="proxy",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="200",le="0.5"} 2
grafana_alerting_request_duration_seconds_bucket{backend="proxy",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="200",le="1"} 2
grafana_alerting_request_duration_seconds_bucket{backend="proxy",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="200",le="2.5"} 2
grafana_alerting_request_duration_seconds_bucket{backend="proxy",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="200",le="5"} 2
grafana_alerting_request_duration_seconds_bucket{backend="proxy",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="200",le="10"} 2
grafana_alerting_request_duration_seconds_bucket{backend="proxy",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="200",le="+Inf"} 2
grafana_alerting_request_duration_seconds_sum{backend="proxy",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="200"} 0.588846348
grafana_alerting_request_duration_seconds_count{backend="proxy",method="GET",route="api_ruler_recipient_api_v1_rules",status_code="200"} 2
```

### CODE REVIEW

**Comment by owen-d** [comment]
Note: It'd be nicer to implement at a higher level of middleware, but it was difficult to figure out how to do this. I'm open to suggestions or improving it later.

File: pkg/services/ngalert/api/metrics.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+			Namespace: "grafana",
+			Subsystem: "alerting",
+			Name:      "silences",
+			Help:      "The total number of silences by state.",
+		}, []string{"state"}),
+	}
+}
+
+// Instrument wraps a middleware, instrumenting the request latencies.
+func Instrument(
```
Comment (gotjosh) [inline pkg/services/ngalert/api/metrics.go:93]: Looks 👌 ✨ 

If you wanted to go with a middleware-based approach we can add this to https://github.com/grafana/grafana/blob/master/pkg/api/routing/route_register.go

File: pkg/services/ngalert/api/metrics.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+		start := time.Now()
+		var res response.Response
+		val, err := c.Invoke(action)
+		if err == nil && val != nil && len(val) > 0 {
+			res = val[0].Interface().(response.Response)
+		} else {
+			res = routing.ServerError(err)
+		}
+
+		// TODO: We could look up the datasource type via our datasource service
```
Comment (gotjosh) [inline pkg/services/ngalert/api/metrics.go:111]: I don't think we should/need to do this - we don't have access to the datasource directly via the URL IIRC.

**Review by gotjosh — APPROVED** [review]
LGTM 🌟 

I pretty much like the fact that we've moved the complexity to the generated code. If you wanted to opt for the middleware approach, using `Instrument` on the `RouteRegister` is an option.

**Pull request merged**